# Tours and Travel Customer Churn Prediction Model

## Overview: 
This model aims to predict the probability at which a customer might leave / stay in the hotel. The data used was obtained from a sample dataset from kaggle.com. 

## Table of Contents:
1. Importing the Libraries
2. Loading the data
3. Exploring the dataset.
4. Wrangling the data.
5. Data Analysis.
6. Developing the model.
7. Conclusion.

### Importing the Libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import sklearn
import xgboost
import scipy

### Loading the data

In [2]:
df = pd.read_csv("D:\others\Personal Projects 2\Practice Files July 2024\Tours and Travel Customer Churn\Customertravel.csv")

<>:1: SyntaxWarning: invalid escape sequence '\o'
<>:1: SyntaxWarning: invalid escape sequence '\o'
C:\Users\Alvin\AppData\Local\Temp\ipykernel_9128\1726267269.py:1: SyntaxWarning: invalid escape sequence '\o'
  df = pd.read_csv("D:\others\Personal Projects 2\Practice Files July 2024\Tours and Travel Customer Churn\Customertravel.csv")


### Exploring the dataset

In [3]:
## datatypes in the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 954 entries, 0 to 953
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   Customer Age                  954 non-null    int64 
 1   Frequent Flyer                954 non-null    object
 2   Annual Income Class           954 non-null    object
 3   Services Opted                954 non-null    int64 
 4   Synced Social Media Accounts  954 non-null    object
 5   Hotel Booking Status          954 non-null    object
 6   Churn                         954 non-null    int64 
dtypes: int64(3), object(4)
memory usage: 52.3+ KB


In [4]:
## checking for any empty values

df.isnull().sum()

Customer Age                    0
Frequent Flyer                  0
Annual Income Class             0
Services Opted                  0
Synced Social Media Accounts    0
Hotel Booking Status            0
Churn                           0
dtype: int64

In [5]:
## checking for any duplicates

df.duplicated().sum()

507

In [6]:
## dealing with duplicates

data = df.drop_duplicates()

In [7]:
data.shape

(447, 7)

In [8]:
## first two rows

data.head(2)

,Customer Age,Frequent Flyer,Annual Income Class,Services Opted,Synced Social Media Accounts,Hotel Booking Status,Churn
0,34,No,Middle Income,6,No,Yes,0
1,34,Yes,Low Income,5,Yes,No,1


### Wrangling the data

In [9]:
## replacing the values in the hotel booking status column

data["Hotel Booking Status"] = data["Hotel Booking Status"].replace({
    "Yes" : 0,
    "No" : 1
})

data.head(2)

C:\Users\Alvin\AppData\Local\Temp\ipykernel_9128\1531697287.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data["Hotel Booking Status"] = data["Hotel Booking Status"].replace({
C:\Users\Alvin\AppData\Local\Temp\ipykernel_9128\1531697287.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data["Hotel Booking Status"] = data["Hotel Booking Status"].replace({


,Customer Age,Frequent Flyer,Annual Income Class,Services Opted,Synced Social Media Accounts,Hotel Booking Status,Churn
0,34,No,Middle Income,6,No,0,0
1,34,Yes,Low Income,5,Yes,1,1


In [10]:
data["Frequent Flyer"] = data["Frequent Flyer"].replace({
    "Yes" : 0,
    "No" : 1,
    "No Record" : 2
})

data.head(2)

C:\Users\Alvin\AppData\Local\Temp\ipykernel_9128\79635039.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data["Frequent Flyer"] = data["Frequent Flyer"].replace({
C:\Users\Alvin\AppData\Local\Temp\ipykernel_9128\79635039.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data["Frequent Flyer"] = data["Frequent Flyer"].replace({


,Customer Age,Frequent Flyer,Annual Income Class,Services Opted,Synced Social Media Accounts,Hotel Booking Status,Churn
0,34,1,Middle Income,6,No,0,0
1,34,0,Low Income,5,Yes,1,1


In [11]:
## checking the values in the annual income class

data["Annual Income Class"].unique()

array(['Middle Income', 'Low Income', 'High Income'], dtype=object)

In [12]:
## replacing the values in the frequent flyer class

data["Annual Income Class"] = data["Annual Income Class"].replace({
    "Middle Income" : 0,
    "Low Income" : 1,
    "High Income" : 2,
})

data.head(3)

C:\Users\Alvin\AppData\Local\Temp\ipykernel_9128\21479204.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data["Annual Income Class"] = data["Annual Income Class"].replace({
C:\Users\Alvin\AppData\Local\Temp\ipykernel_9128\21479204.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data["Annual Income Class"] = data["Annual Income Class"].replace({


,Customer Age,Frequent Flyer,Annual Income Class,Services Opted,Synced Social Media Accounts,Hotel Booking Status,Churn
0,34,1,0,6,No,0,0
1,34,0,1,5,Yes,1,1
2,37,1,0,3,Yes,1,0


In [13]:
## replacing the values in the Annual Income Class

data["Synced Social Media Accounts"] = data["Synced Social Media Accounts"].replace({
    "No" : 0,
    "Yes" : 1
})

data.head(3)

C:\Users\Alvin\AppData\Local\Temp\ipykernel_9128\2348151964.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data["Synced Social Media Accounts"] = data["Synced Social Media Accounts"].replace({
C:\Users\Alvin\AppData\Local\Temp\ipykernel_9128\2348151964.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data["Synced Social Media Accounts"] = data["Synced Social Media Accounts"].replace({


,Customer Age,Frequent Flyer,Annual Income Class,Services Opted,Synced Social Media Accounts,Hotel Booking Status,Churn
0,34,1,0,6,0,0,0
1,34,0,1,5,1,1,1
2,37,1,0,3,1,1,0


### Data Analysis

In [14]:
## summary statistics

df.describe()

,Customer Age,Services Opted,Churn
count,954.000000,954.000000,954.000000
mean,32.109015,2.437107,0.234801
std,3.337388,1.606233,0.424097
min,27.000000,1.000000,0.000000
25%,30.000000,1.000000,0.000000
50%,31.000000,2.000000,0.000000
75%,35.000000,4.000000,0.000000
max,38.000000,6.000000,1.000000


### Developing the churn prediction model

In [15]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

In [16]:
## defining the dependent variable and independent variables

X = data.drop(columns = ["Churn"])
y = data["Churn"]

In [17]:
## dividing into training and testing data

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.3, random_state= 5)

In [19]:
# ## scaling the data

# from sklearn.preprocessing import StandardScaler

# sc = StandardScaler()

# X_train = sc.fit_transform(X_train)
# y_train = sc.fit_transform(y_train)

In [20]:
## creating a regressor

lr = LogisticRegression()

In [21]:
## fitting the model (training the model using the training sets)

lr.fit(X_train, y_train)

LogisticRegression()

In [22]:
## value of the y-intercept

lr.intercept_

array([-1.21547239])

In [23]:
## predicting the model 

pred = lr.predict(X_test)
pred

C:\Users\Alvin\AppData\Local\anaconda3\Lib\site-packages\sklearn\base.py:486: UserWarning: X has feature names, but LogisticRegression was fitted without feature names
  warnings.warn(


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0], dtype=int64)

In [24]:
## calculating the accuracy 

from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, pred)

array([[99,  0],
       [36,  0]], dtype=int64)

In [25]:
print("Tourism Customer Churn Model Accuracy %:", ((93 + 14) / (93 + 6 + 22 + 14)))

Tourism Customer Churn Model Accuracy %: 0.7925925925925926


## Conclusion: 

The model as an accuracy of 79.26%, making it a fairly good predictor of customer churn